### ISSUE TITLE
Filter HttpResponse  return two same headers to client

### ISSUE THREAD

**Issue by jthann** [megaease/easegress#480]
**Describe the bug**
In any one filter，as follow example shows setting some response header:
```go
func (f *XXXFilter) Handle(ctx context.HTTPContext) (result string) {
    ctx.Response().Header().Set("key-xxx", "value-xxx")
    ctx.Response().Header().Set("key-yyy", "value-yyy")
}
```
The client(whether using curl or chrome browser) will receive two same headers:
key-xxx: value-xxx
key-xxx: value-xxx
key-yyy: value-yyy
key-yyy: value-yyy

This will cause CORSAdaptor filter not working properly

**Version**
v1.4.1

**Comment by localvar** [comment]
I'm not quite clear with the description, are you developing a new filter and setting headers from it?

**Comment by jthann** [comment]
> I'm not quite clear with the description, are you developing a new filter and setting headers from it?

Yes，I pulled the easegress source code and checkout new branch from tag v1.4.1 using goland IDE
developing a new filter to satisfy our biz requirement，but the problem first happened in easegress builtin filter CORSAdaptor(our api service and web page service have different domain)，later on I test this case in my custom filter，the problem also exists

**Comment by samutamm** [comment]
I tried this kind of dummy filter that sets two **response** headers, in a HTTP pipeline of two filters: `DummyFilter --> Proxy`. Using `curl` I observed that if the Proxy backend is available, the headers are not duplicated. However, if the backend is not available, `curl` returns status 503 as expected and the headers duplicated. I find this interesting, but not sure if it is due Easegress or not (needs more investigation).

> This will cause CORSAdaptor filter not working properly

Can you provide more details on this? What did not work? Was the value of `Access-Control-Allow-Origin` set as you expected?

**Comment by jthann** [comment]
Yes，if backend is not available, curl returns status 503 and the headers duplicated，I think this is Easegress bug, 
If you use builtin filter CORSAdaptor testing, whether or not  backend is available，CORSAdaptor always return duplicated header, this will cause cors request not working and block our normal usage

Using following Easegress objects yaml config 

```yaml
kind: GlobalFilter
name: globalFilter
beforePipeline:
  flow:
    - filter: cors-filter
      jumpIf: { preflighted: END }
    - filter: dummy-filter
  filters:
    - name: cors-filter
      kind: CORSAdaptor
      allowedOrigins: ["http://frontend.local"]
      allowCredentials: true
      allowedMethods: ["GET","POST"]
    - name: dummy-filter
      kind: DummyFilter
---

kind: HTTPServer
name: http
port: 80
keepAlive: true
https: false
globalFilter: globalFilter
rules:
  - host: frontend.local
    paths:
      - PathPrefix: /page
        backend: frontend
  - host: api.backend.local
    paths:
      - pathPrefix: /api
        backend: backend

---

name: frontend
kind: HTTPPipeline
flow:
  - filter: proxy
filters:
  - name: proxy
    kind: Proxy
    mainPool:
      servers:
        - url: http://127.0.0.1:8080
      loadBalance:
        policy: roundRobin

---

name: backend
kind: HTTPPipeline
flow:
  - filter: proxy
filters:
  - name: proxy
    kind: Proxy
    mainPool:
      servers:
        - url: http://127.0.0.1:8081
      loadBalance:
        policy: roundRobin
```

local /etc/hosts config
```
127.0.0.1 frontend.local api.backend.local
```

In frontend app just return a html page，in the html just send a fetch request to backend api
```html
<!DOCTYPE html>
<html>
<head>
    <meta charset="utf-8">
</head>
<body>
<div id="container">
    <h1>This is Home!</h1>
</div>
<script>
    fetch('http://api.backend.local/api/echo', {
        mode: "cors",
        method: "POST",
        credentials: "include",
        cache: 'no-cache',
        headers: {
            'Content-Type': 'application/json'
            // 'Content-Type': 'application/x-www-form-urlencoded',
        },
    }).then(response => response.json())
      .then(data => console.log(data));
</script>
</body>
</html>
```

This will never work normally, and preflight OPTIONS request return duplicated headers
that will lead to later POST request report cors error

**Comment by samutamm** [comment]
@jthann Thanks for detailed test case. You are right, it's a bug. Here's a fix https://github.com/megaease/easegress/pull/482, could you let us know if it fixes the bug.

**Comment by jthann** [comment]
> @jthann Thanks for detailed test case. You are right, it's a bug. Here's a fix #482, could you let us know if it fixes the bug.

Yes，the duplicated response header problem has solved.
But I think CORSAdaptor implementation has some problem and still cannot work，maybe I should open another bug issue.

**Comment by samutamm** [comment]
Yeah let's track it in an another issue.
